[WebXR] Opaque WebGLFramebuffer incorrect behavior

djg opened the pull request:
#### b4d954c0185e8130346ab41ca9180424496f9b29
<pre>
[WebXR] Opaque WebGLFramebuffer incorrect behavior
<a href="https://bugs.webkit.org/show_bug.cgi?id=301407">https://bugs.webkit.org/show_bug.cgi?id=301407</a>
<a href="https://rdar.apple.com/163317425">rdar://163317425</a>

Reviewed by Kimmo Kinnunen.

xrWebGLLayer_opaque_framebuffer.https.html fails due to two oversights:
- WebGL2RenderingContext::deleteFramebuffer override is missing the opaque
  check from WebGLRenderingContextBase::deleteFramebuffer.
- WebGLRenderingContextBase::checkFramebufferStatus does not return
  FRAMEBUFFER_UNSUPPORTED when an opaque framebuffer is accessed outside
  a requestAnimationFrame() callback, as required by the WebXR specification.

This change fixes the test by:
1. Adding the opaque framebuffer validation check to
   WebGL2RenderingContext::deleteFramebuffer (matching the base class).
2. Updating WebGLRenderingContextBase::checkFramebufferStatus to return
   FRAMEBUFFER_UNSUPPORTED when an opaque framebuffer is accessed outside
   a WebXR requestAnimationFrame() callback.
3. Tracking WebXR rAF state via WebGLFramebuffer::setInsideWebXRRAF(),
   called from WebXROpaqueFramebuffer::startFrame() and endFrame().

Change expectations and enable xrWebGLLayer_opaque_framebuffer.https.html

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/webxr/xrWebGLLayer_opaque_framebuffer.https-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/visionos/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::deleteFramebuffer):
* Source/WebCore/html/canvas/WebGLFramebuffer.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::checkFramebufferStatus):
(WebCore::WebGLRenderingContextBase::deleteFramebuffer):

Canonical link: <a href="https://commits.webkit.org/302324@main">https://commits.webkit.org/302324@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a09c9d1bbf241e60ded784e606878a4e6727fe4c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128764 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1021 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39594 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136146 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/80140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ad998ad3-22f2-4f77-bbe6-e355df97dca3) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130635 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/972 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/898 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/98024 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/80140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e06d6e67-7e34-4c12-98d5-c7d1416a0293) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131711 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/723 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/115347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78632 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6dc2988e-9cd9-4ace-a0cb-52780f17d964) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/33460 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79426 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109093 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/33954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138605 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/838 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/815 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/106564 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/891 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/111685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/106372 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/27079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/30216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53251 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20112 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/907 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64195 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/849 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -1203,6 +1209,18 @@ GCGLenum WebGLRenderingContextBase::checkFramebufferStatus(GCGLenum target)
+    RefPtr framebuffer = getFramebufferBinding(target);
+    if (framebuffer && framebuffer->isOpaque() && !framebuffer->isInsideWebXRRAF()) {
+        synthesizeGLError(GraphicsContextGL::FRAMEBUFFER_UNSUPPORTED, "checkFramebufferStatus"_s, "called outside WebXR requestAnimationFrame() callback"_s);
+        return GraphicsContextGL::FRAMEBUFFER_UNSUPPORTED;
```
when checkFramebufferStatus runs into an error, like in the case above, the return value should be zero.
 when checkFrameBufferStatus result is FRAMEBUFFER_UNSUPPORTED, it should not generate an error. It's business as usual, result.
 The spec wording is wrong, FRAMEBUFFER_UNSUPPORTED is not an error.
 (So remove the synthesizeGLError.

kkinnunen-apple commented on Source/WebCore/platform/graphics/GraphicsContextGL.h:
```diff
@@ -1225,6 +1225,8 @@ class GraphicsContextGL : public RefCounted<GraphicsContextGL> {
             return GCGLErrorCode::InvalidFramebufferOperation;
         case CONTEXT_LOST_WEBGL:
             return GCGLErrorCode::ContextLost;
+        case FRAMEBUFFER_UNSUPPORTED:
```
Can't be added here, FRAMEBUFFER_UNSUPPORTED is not an error.

kkinnunen-apple reviewed (commented):
something to fix

svillar commented on LayoutTests/platform/gtk/TestExpectations:
```diff
@@ -98,6 +98,9 @@ webkit.org/b/254733 inspector/canvas/shaderProgram-add-remove-webgl.html [ Failu
 
+# WebXR
+webkit.org/b/301551 imported/w3c/web-platform-tests/webxr/xrWebGLLayer_opaque_framebuffer.https.html [ Failure ]
+
```
This is surprising. According to the bug it's failing with an incomplete attachment error. But the code is the same as the one used in vision which behaves as expected...

djg replied:
Does GTK use the ANGLE GraphicsContext backend?

svillar replied:
It uses a subclass of the ANGLE GC.

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLFramebuffer.h:
```diff
@@ -119,12 +128,18 @@ class WebGLFramebuffer final : public WebGLObject {
     Vector<GCGLenum> m_filteredDrawBuffers;
 #if ENABLE(WEBXR)
     const bool m_isOpaque;
+    bool m_insideWebXRRAF = false;
```
I think the initialiser syntax is with braces: `{ false }`

djg committed:
[WebXR] Opaque WebGLFramebuffer incorrect behavior
https://bugs.webkit.org/show_bug.cgi?id=301407
rdar://163317425

Reviewed by Kimmo Kinnunen.

xrWebGLLayer_opaque_framebuffer.https.html fails due to two oversights:
- WebGL2RenderingContext::deleteFramebuffer override is missing the opaque
  check from WebGLRenderingContextBase::deleteFramebuffer.
- WebGLRenderingContextBase::checkFramebufferStatus does not return
  FRAMEBUFFER_UNSUPPORTED when an opaque framebuffer is accessed outside
  a requestAnimationFrame() callback, as required by the WebXR specification.

This change fixes the test by:
1. Adding the opaque framebuffer validation check to
   WebGL2RenderingContext::deleteFramebuffer (matching the base class).
2. Updating WebGLRenderingContextBase::checkFramebufferStatus to return
   FRAMEBUFFER_UNSUPPORTED when an opaque framebuffer is accessed outside
   a WebXR requestAnimationFrame() callback.
3. Tracking WebXR rAF state via WebGLFramebuffer::setInsideWebXRRAF(),
   called from WebXROpaqueFramebuffer::startFrame() and endFrame().

Change expectations and enable xrWebGLLayer_opaque_framebuffer.https.html

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/webxr/xrWebGLLayer_opaque_framebuffer.https-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/visionos/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::deleteFramebuffer):
* Source/WebCore/html/canvas/WebGLFramebuffer.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::checkFramebufferStatus):
(WebCore::WebGLRenderingContextBase::deleteFramebuffer):

Canonical link: https://commits.webkit.org/302324@main
